src/main/scala/com/databricks/spark/csv/CsvRelation.scala

```diff
@@ -136,6 +136,58 @@ case class CsvRelation protected[spark] (
     }
   }
 
+  override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
+    val schemaFields = schema.fields
+    val isTableScan = requiredColumns.isEmpty || requiredColumns.sameElements(schemaFields)
+    if (isTableScan) {
+      buildScan
+    } else {
+      val requiredFields = StructType(requiredColumns.map(schema(_))).fields
+      val requiredIndices = new Array[Int](requiredFields.length)
+      schemaFields.zipWithIndex.filter(pair => requiredFields.contains(pair._1))
+        .foreach(pair => requiredIndices.update(requiredFields.indexOf(pair._1), pair._2))
+
+      tokenRdd(schemaFields.map(_.name)).flatMap { tokens =>
+        if (dropMalformed && schemaFields.length != tokens.size) {
+          logger.warn(s"Dropping malformed line: ${tokens.mkString(",")}")
+          None
+        } else if (failFast && schemaFields.length != tokens.size) {
+          throw new RuntimeException(s"Malformed line in FAILFAST mode: ${tokens.mkString(",")}")
+        } else {
+          val indexSafeTokens = if (permissive && schemaFields.length != tokens.size) {
+            tokens ++ new Array[String](schemaFields.length - tokens.length)
+          } else {
+            tokens
+          }
+          var index: Int = 0
+          var subIndex: Int = 0
+          val rowArray = new Array[Any](requiredIndices.length)
```

Comment:
If you're interested in leveraging mutability as a performance optimization, then I think you might be able to re-use a single mutable array rather than creating a new one for each row.